### PR TITLE
Make server more robust in production

### PR DESCRIPTION
**Pull request by cpcallen**
Should solve most of the problems that result when disk-full or other problems prevent successful checkpoints.  See individual changes for details.

### CODE REVIEW

File: static/503.html [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      display: block;
+      margin: 0 auto;
+    }
+  </style>
+  <link href="https://fonts.googleapis.com/css?family=Roboto+Mono" rel="stylesheet">
+  <link href="/static/favicon.ico" rel="shortcut icon">
+</head>
+<body>
+  <h1>
+    <svg xmlns="http://www.w3.org/2000/svg" width="95px" height="100px">
```
Comment (NeilFraser) [inline static/503.html]: If you approve/merge PR #116, then this \<svg\> can be replaced with:
`    `
Also edit `h1>svg` in the CSS above to be `h1>img`

File: static/503.html [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      font-weight: 100;
+    }
+    h1>svg {
+      vertical-align: text-bottom;
+    }
+    #tagline {
+      font-style: italic;
+      margin: 2em;
+    }
+    iframe {
```
Comment (NeilFraser) [inline static/503.html]: CSS rule for iframes may be deleted.

File: static/503.html [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<!doctype html>
+<html lang="en">
+<head>
+  <title>Code City</title>
```
Comment (NeilFraser) [inline static/503.html]: 503 - Code City
(to match the existing "404 - Code City" page)

File: static/503.html [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    <svg xmlns="http://www.w3.org/2000/svg" width="95px" height="100px">
+      <path
+         d="m 1,90 0,-25 10,0 0,-8 9,0 0,-5 8,0 0,5 2,0 0,33 3,0 0,-13 2,0 0,-7 2,0 0,-32 1,0 0,-7 1,0 0,-6 2,0 0,-1 1,0 0,-12 1,0 0,-12 0,12 1,0 0,12 1,0 0,1 2,0 0,6 1,0 0,7 1,0 0,32 2,0 0,7 2,0 0,13 4,0 0,-32 5,0 0,-5 6,0 0,5 3,0 0,32 5,0 0,-42 3,0 0,-9 7,0 0,-2 5,0 0,2 2,0 0,51"
+         style="fill: none; stroke: #000; stroke-width: 1.5;" />
+    </svg>
+    Code City
+  </h1>
+  <p id="tagline">A community of inquisitive programmers.</p>
+  <p>Unfortunately this Code City is not available right now. 
+  Please come back later!</p>
```
Comment (NeilFraser) [inline static/503.html]: The target audience are programmers.  Can we create a less corporate-like message?  In this context it's ok to say we screwed up, it's ok to have fun, and it's ok to divulge any available technical details.

File: server/codecity.js [hunk 5]
```diff
@@ -150,15 +150,36 @@ CodeCity.checkpoint = function(sync) {
 
   var filename = (new Date()).toISOString().replace(/:/g, '.') + '.city';
   filename = path.join(CodeCity.databaseDirectory, filename);
-  fs.writeFileSync(filename, text);
+  var tmpFilename = filename + '.partial';
+  try {
+    fs.writeFileSync(tmpFilename, text);
+    fs.renameSync(tmpFilename, filename);
+    console.log('Checkpoint complete.');
+  } catch(e) {
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/codecity.js]: Done.

File: server/codecity.js [hunk 6]
```diff
@@ -150,15 +150,36 @@ CodeCity.checkpoint = function(sync) {
...
+    console.error('Checkpoint failed!  ' + e);
+  } finally {
+    // Attempt to remove partially-written checkpoint if it still exists.
+    try {
+      fs.unlinkSync(tmpFilename);
+    } catch(e) {
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/codecity.js]: Done.

File: static/503.html (hunk 2 above)
Comment (cpcallen) [reply static/503.html]: Done.

File: static/503.html (hunk 1 above)
Comment (cpcallen) [reply static/503.html]: Done.

File: static/503.html (hunk 3 above)
Comment (cpcallen) [reply static/503.html]: Done.  I made some other changes for consistency as well.  Previously I had tried to make the 503 page look like the login page, but now it is much more like the 404 page.

File: static/503.html (hunk 4 above)
Comment (cpcallen) [reply static/503.html]: Yes.  I've rewritten it to provide a bit more detail about possible causes (including incompetence on our part).  Do improve it further if you have other ideas.

**Pull request merged**